source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.h

```diff
@@ -1,11 +1,234 @@
+        if (shutting_down_) {
+          break;
+        }
+
+        next_query = std::move(pending_queries_.front());
+        pending_queries_.pop_front();
+        if (reresolve && next_query->cancelled_) {
+          continue;
+        }
+      }
+
+      ENVOY_LOG(debug, "popped pending query [{}]", next_query->dns_name_);
+
+      // For mock testing make sure the getaddrinfo() response is freed prior to the post.
+      std::pair<ResolutionStatus, std::list<DnsResponse>> response;
+      {
+        addrinfo hints;
+        memset(&hints, 0, sizeof(hints));
+        hints.ai_flags = AI_ADDRCONFIG;
+        hints.ai_family = AF_UNSPEC;
+        // If we don't specify a socket type, every address will appear twice, once
+        // for SOCK_STREAM and one for SOCK_DGRAM. Since we do not return the family
+        // anyway, just pick one.
+        hints.ai_socktype = SOCK_STREAM;
+        addrinfo* addrinfo_result_do_not_use = nullptr;
+        auto rc = Api::OsSysCallsSingleton::get().getaddrinfo(
+            next_query->dns_name_.c_str(), nullptr, &hints, &addrinfo_result_do_not_use);
+        auto addrinfo_wrapper = AddrInfoWrapper(addrinfo_result_do_not_use);
+        if (rc.return_value_ == 0) {
+          response = processResponse(*next_query, addrinfo_wrapper.get());
+        } else if (reresolve && rc.return_value_ == EAI_AGAIN) {
+          ENVOY_LOG(debug, "retrying query [{}]", next_query->dns_name_);
+          absl::MutexLock guard(&mutex_);
+          pending_queries_.push_back(next_query);
+          continue;
+        } else {
+          // TODO(mattklein123): Handle some errors differently such as `EAI_NODATA`.
+          ENVOY_LOG(debug, "getaddrinfo failed with rc={} errno={}", gai_strerror(rc.return_value_),
+                    errorDetails(rc.errno_));
+          response = std::make_pair(ResolutionStatus::Failure, std::list<DnsResponse>());
+        }
+      }
+
+      dispatcher_.post(
+          [finished_query = std::move(next_query), response = std::move(response)]() mutable {
+            if (finished_query->cancelled_) {
+              ENVOY_LOG(debug, "dropping cancelled query [{}]", finished_query->dns_name_);
+            } else {
+              finished_query->callback_(response.first, std::move(response.second));
+            }
+          });
+    }
+
+    ENVOY_LOG(debug, "getaddrinfo resolver thread exiting");
+  }
+
+  // getaddrinfo() doesn't provide TTL so use a hard coded default. This can be made configurable
+  // later if needed.
+  static constexpr std::chrono::seconds DEFAULT_TTL = std::chrono::seconds(60);
+
+  Event::Dispatcher& dispatcher_;
+  absl::Mutex mutex_;
+  std::list<PendingQuerySharedPtr> pending_queries_ ABSL_GUARDED_BY(mutex_);
+  bool shutting_down_ ABSL_GUARDED_BY(mutex_){};
+  // The resolver thread must be initialized last so that the above members are already fully
+  // initialized.
+  const Thread::ThreadPtr resolver_thread_;
+};
+
```

Comment:
Hm. There is a lot of code that is now in the .h file which is slightly unfortunate. Could we keep the definitions in the .cc file? Since this PR is 100% move, feel free to do in a follow-up, if you agree.

Reply:
done